### PR TITLE
Kernel: Add FATFS write support

### PR DESCRIPTION
**Pull request by implicitfield**
This continues where #22852 left off, so the majority of the commits here are based on those in that PR. I've tested the write support on all the FAT variants we support, and `fsck_msdos` on macOS comes up with no errors when checking a filesystem that has been modified by this driver.

### CODE REVIEW

File: Kernel/FileSystem/FATFS/Inode.cpp [hunk 1]
```diff
@@ -57,7 +57,7 @@ ErrorOr<Vector<BlockBasedFileSystem::BlockIndex>> FATInode::compute_block_list()
 
         auto first_block_and_length = fs().first_block_of_cluster(cluster);
         for (u8 i = 0; i < first_block_and_length.number_of_sectors; i++)
-            block_list.append(BlockBasedFileSystem::BlockIndex { first_block_and_length.start_block.value() + i });
+            TRY(block_list.try_append(BlockBasedFileSystem::BlockIndex { first_block_and_length.start_block.value() + i }));
```
Comment (Hendiadyoin1) [inline Kernel/FileSystem/FATFS/Inode.cpp]: Using `Better` as a verb is stretching it, but I'll allow it
(`Kernel/FATFS: Better error propagation in FATInode`)

**Review by cqundefine — COMMENTED** [review]
One small thing I noticed

File: Kernel/FileSystem/FATFS/Inode.cpp [hunk 2]
```diff
@@ -602,9 +602,64 @@ ErrorOr<void> FATInode::add_child(Inode& inode, StringView name, mode_t mode)
...
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    Vector<FATLongFileNameEntry> lfn_entries;
+    TRY(lfn_entries.try_ensure_capacity(ceil_div(max_filename_length, characters_per_lfn_entry)));
+
+    Vector<FATEntryLocation> lfn_entry_locations;
+    TRY(lfn_entry_locations.try_ensure_capacity(ceil_div(max_filename_length, characters_per_lfn_entry)));
+
+    auto block_list = TRY(get_block_list());
+    auto blocks = TRY(read_block_list());
```
Comment (timschumi) [inline Kernel/FileSystem/FATFS/Inode.cpp:644]: Maybe one of these two could use a better name.

File: Kernel/FileSystem/FATFS/Inode.cpp [hunk 3]
```diff
@@ -555,6 +555,29 @@ ErrorOr<NonnullRefPtr<Inode>> FATInode::create_child(StringView name, mode_t mod
...
+            memset(directory_entry.filename, ' ', 8);
+            memset(directory_entry.extension, ' ', 3);
+            for (size_t i = 0; i < entry_name.length(); ++i)
+                directory_entry.filename[i] = entry_name[i];
+            directory_entry.attributes |= FATAttributes::Directory;
+            return directory_entry;
+        };
+
+        FATEntry current_directory = create_directory_entry("."sv);
+        current_directory.first_cluster_high = entry.first_cluster_high;
```
Comment (timschumi) [inline Kernel/FileSystem/FATFS/Inode.cpp:572]: Does this need to be guarded with a FAT32 check?

File: Kernel/FileSystem/FATFS/Inode.cpp [hunk 4]
```diff
@@ -676,6 +676,23 @@ ErrorOr<void> FATInode::remove_child(StringView name)
                 for (auto const& lfn_entry_location : lfn_entry_locations)
                     TRY(fs().write_block(lfn_entry_location.block, UserOrKernelBuffer::for_kernel_buffer(bit_cast<u8*>(&unused_entry)), sizeof(FATEntry), lfn_entry_location.entry * sizeof(FATEntry)));
 
+                u32 entry_first_cluster = entry->first_cluster_low;
+                if (fs().m_fat_version == FATVersion::FAT32)
+                    entry_first_cluster = (static_cast<u32>(entry->first_cluster_high) << 16) | entry->first_cluster_low;
```
Comment (timschumi) [inline Kernel/FileSystem/FATFS/Inode.cpp:681]:
```suggestion
                    entry_first_cluster |= (static_cast<u32>(entry->first_cluster_high) << 16);
```

**Review by timschumi — CHANGES_REQUESTED** [review]
Looking good so far, mainly a bunch of smaller nitpicks.

File: Kernel/FileSystem/FATFS/FileSystem.cpp [hunk 5]
```diff
@@ -246,6 +246,24 @@ ErrorOr<void> FATFS::initialize_while_locked()
...
+        auto fs_info_buffer = UserOrKernelBuffer::for_kernel_buffer(bit_cast<u8*>(&m_fs_info));
+        // We know that there is a DOS7 BPB, because if it wasn't present
+        // we would have returned EINVAL above.
+        TRY(read_block(ebpb.dos7_bpb()->fs_info_sector, &fs_info_buffer, sizeof(m_fs_info)));
+
+        if (m_fs_info.lead_signature != fs_info_signature_1 || m_fs_info.struct_signature != fs_info_signature_2 || m_fs_info.trailing_signature != fs_info_signature_3) {
+            dbgln("FATFS: Invalid FSInfo struct signature");
+            dbgln_if(FAT_DEBUG, "FATFS: FSInfo signature1: {}, expected: {}", m_fs_info.lead_signature, fs_info_signature_1);
+            dbgln_if(FAT_DEBUG, "FATFS: FSInfo signature2: {}, expected: {}", m_fs_info.struct_signature, fs_info_signature_2);
+            dbgln_if(FAT_DEBUG, "FATFS: FSInfo signature3: {}, expected: {}", m_fs_info.trailing_signature, fs_info_signature_3);
```
[the remark this answers is not in the document]
Comment (nico) [reply Kernel/FileSystem/FATFS/FileSystem.cpp:259]: (Use `{:#x}` instead of `{}` to do that.)

**Review by timschumi — COMMENTED** [review]
Preliminary "LGTM", I'll give it a test run tomorrow. If you want to add your "Co-Authored-By" anywhere, now is the time. :^)

**Comment by timschumi** [comment]
FAT images (all types) are exhibiting weird behavior for me now, even without writing anything. The attached images ([fat60.zip](https://github.com/SerenityOS/serenity/files/15049249/fat60.zip)) were created using `mkfs.fat` and populated under Linux, but stall very for very long while working on them under Serenity (FAT12 even when just listing the root directory, the others when trying to read the large bin file).

**Review by timschumi — APPROVED** [review]
Still a bit on the slow side with the new commits, but at least it can be worked with. :^)

**Pull request merged**